File: src/gm.c
#include <stdio.h>
#include <math.h>
#include "gm.h"

vec2 new_vec2(f32 x, f32 y) {
	return (vec2) {
		x,
		y,
	};
}

vec3 new_vec3(f32 x, f32 y, f32 z) {
	return (vec3) {
		x,
		y,
		z,
	};
}

vec4 new_vec4(f32 x, f32 y, f32 z, f32 w) {
	return (vec4) {
		x,
		y,
		z,
		w,
	};
}

vertex new_vertex(vec3 pos, vec2 tex_coord) {
	return (vertex){
		pos,
		tex_coord,
	};
}

quad new_quad(vertex tl, vertex tr, vertex bl, vertex br) {
	return (quad){
		tl,
		tr,
		bl,
		br,
	};
}

rect new_rect(f32 x, f32 y, f32 w, f32 h) {
	return (rect){
		.x = x,
		.y = y,
		.width = w,
		.height = h,
	};
}

mat4 identity() {
	return (mat4){
		.data = {
			1, 0, 0, 0,
			0, 1, 0, 0,
			0, 0, 1, 0,
			0, 0, 0, 1,
		},
	};
}

mat4 ortho_proj(f32 l, f32 r, f32 t, f32 b, f32 f, f32 n) {
	return (mat4){
		.data = {
			2/(r-l),      0,            0,            0,
			0,            2/(t-b),      0,            0,
			0,            0,            -2/(f-n),     0,
			-(r+l)/(r-l), -(t+b)/(t-b), -(f+n)/(f-n), 1,
		}
	};
}

vec2 zero2() {
	return (vec2) {
		0.0f,
		0.0f,
	};
}

vec3 zero3() {
	return (vec3) {
		0.0f,
		0.0f,
		0.0f,
	};
}

vec4 zero4() {
	return (vec4) {
		0.0f,
		0.0f,
		0.0f,
		0.0f,
	};
}

vec3 swizzle3f2(vec2 vec) {
	return (vec3){
		vec.x,
		vec.y,
		1.0f,
	};
}

vec4 swizzle4f2(vec2 vec) {
	return (vec4) {
		vec.x,
		vec.y,
		1.0f,
		1.0f,
	};
}

vec4 swizzle4f3(vec3 vec) {
	return (vec4) {
		vec.x,
		vec.y,
		vec.z,
		1.0f,
	};
}

vec3 add_vec3(vec3 left, vec3 right) {
	left.x += right.x;
	left.y += right.y;
	left.z += right.z;

	return left;
}

vec3 sub_vec3(vec3 left, vec3 right) {
	left.x -= right.x;
	left.y -= right.y;
	left.z -= right.z;

	return left;
}

static f32 y_to_texture_space(f32 y, f32 height) {
	return 1-(y/height);
}

static f32 x_to_texture_space(f32 x, f32 width) {
	return x/width;
}

vec2 vec2_to_texture_space(vec2 v, f32 width, f32 height) {
	v.x = x_to_texture_space(v.x, width);
	v.y = y_to_texture_space(v.y, height);

	return v;
}

bool compare_f32(f32 left, f32 right) {
	return left > (right - 0.00001) && left < (right + 0.00001);
}

f32 mag2(vec2 v) {
	return sqrt(v.x * v.x + v.y * v.y);
}

f32 mag3(vec3 v) {
	return sqrt(v.x * v.x + v.y * v.y + v.z * v.z);
}

f32 mag4(vec4 v) {
	return sqrt(v.x * v.x + v.y * v.y + v.z * v.z + v.w * v.w);
}

vec2 unit2(vec2 v) {
	f32 mag = mag2(v);
	if (compare_f32(mag, 0.0f)) {
		return zero2();
	}

	return new_vec2(v.x/mag, v.y/mag);
}

vec3 unit3(vec3 v) {
	f32 mag = mag3(v);
	if (compare_f32(mag, 0.0f)) {
		return zero3();
	}

	return new_vec3(v.x/mag, v.y/mag, v.z/mag);
}

vec4 unit4(vec4 v) {
	f32 mag = mag4(v);
	if (compare_f32(mag, 0.0f)) {
		return zero4();
	}

	return new_vec4(v.x/mag, v.y/mag, v.z/mag, v.w/mag);
}

vec2 scale2(vec2 v, f32 s) {
	return new_vec2(v.x * s, v.y * s);
}

vec3 scale3(vec3 v, f32 s) {
	return new_vec3(v.x * s, v.y * s, v.z * s);
}

vec4 scale4(vec4 v, f32 s) {
	return new_vec4(v.x * s, v.y * s, v.z * s, v.w * s);
}

/* Column major matrix
 * -------------------
 * 0, 4, 8, 12
 * 1, 5, 9, 13,
 * 2, 6, 10, 14,
 * 3, 7, 11, 15,
*/
mat4 mult_mat4(mat4 l, mat4 r) {
	mat4 res;
	for (i32 i = 0; i < 4; i++) {
		for (i32 j = 0; j < 4; j++) {
			res.data[i+(4*j)] =
				(l.data[i] * r.data[j*4]) +
				(l.data[(i+4)] * r.data[(j*4)+1]) +
				(l.data[(i+8)] * r.data[(j*4)+2]) +
				(l.data[(i+12)] * r.data[(j*4)+3]);
		}
	}

	return res;
}

mat4 translation(vec3 translate) {
	mat4 m = identity();
	m.data[12] = translate.x;
	m.data[13] = translate.y;
	m.data[14] = translate.z;

	return m;
}

mat4 translate_mat4(mat4 m, vec3 translate) {
	mat4 t = translation(translate);
	return mult_mat4(m, t);
}

static bool _overlaps(rect l, rect r) {
	bool y_overlap = (
		(l.y >= r.y && l.y <= (r.y + r.height)) ||
		((l.y + l.height) >= r.y && (l.y + l.height) <= (r.y + r.height))
	);

	bool x_overlap = (
		(l.x <= (r.x + r.width) && l.x >= r.x) ||
		((l.x + l.width) <= (r.x + r.width) && ((l.x + l.width) >= r.x))
	);

	return y_overlap && x_overlap;
}

bool overlaps(rect l, rect r) {
	// flipping the args should detect containment
	return  _overlaps(l, r) || _overlaps(r, l);
}

void translate_rect(rect *r, vec3 v) {
	r->x = v.x;
	r->y = v.y;
}

void print_vec2(vec2 vec) {
	printf("{ x: %.3f, y: %.3f }\n", vec.x, vec.y);
}

void print_vec3(vec3 vec) {
	printf("{ x: %.3f, y: %.3f, z: %.3f }\n", vec.x, vec.y, vec.z);
}

void print_vec4(vec4 vec) {
	printf("{ x: %.3f, y: %.3f, z: %.3f, w: %.3f }\n", vec.x, vec.y, vec.z, vec.w);
}

void print_vertex(vertex vert) {
	printf("Vertex { ");
	print_vec3(vert.pos);
	printf(", ");
	print_vec2(vert.tex_coord);
	printf(" }\n");
}

void print_quad(quad q) {
	printf("Quad { ");
	print_vertex(q.tl);
	print_vertex(q.tr);
	print_vertex(q.bl);
	print_vertex(q.br);
	printf(" }");
	printf("\n");
}

File: rust/src/gm.rs
pub struct Vec2 {
    x: f32,
    y: f32,
}

pub struct Vec3 {
    x: f32,
    y: f32,
    z: f32,
}

pub struct Vec4 {
    x: f32,
    y: f32,
    z: f32,
    w: f32,
}

// Mat4 is stored in column major format.
pub struct Mat4 {
    data: [f32; 16],
}

pub struct Rect {
    pos: Vec2,
    w: f32,
    h: f32,
}

pub struct Vertex {
    pos: Vec3,
    tex_coord: Vec2,
}

pub struct Quad {
    tl: Vertex,
    tr: Vertex,
    bl: Vertex,
    br: Vertex,
}

impl Vec2 {
    pub fn new(x: f32, y: f32) -> Vec2 {
        Vec2 { x, y }
    }

    pub fn zero() -> Vec2 {
        Vec2 { x: 0.0, y: 0.0 }
    }

    pub fn mag(&self) -> f32 {
        (self.x.powi(2) + self.x.powi(2)).sqrt()
    }

    pub fn unit(&self) -> Vec2 {
        let mag = self.mag();
        if f32_eq(mag, 0.0) {
            return Vec2::zero();
        }

        Vec2::new(self.x / mag, self.y / mag)
    }
}

impl Vec3 {
    pub fn new(x: f32, y: f32, z: f32) -> Vec3 {
        Vec3 { x, y, z }
    }

    pub fn zero() -> Vec3 {
        Vec3 {
            x: 0.0,
            y: 0.0,
            z: 0.0,
        }
    }

    pub fn mag(&self) -> f32 {
        (self.x.powi(2) + self.x.powi(2) + self.z.powi(2)).sqrt()
    }

    pub fn unit(&self) -> Vec3 {
        let mag = self.mag();
        if f32_eq(mag, 0.0) {
            return Vec3::zero();
        }

        Vec3::new(self.x / mag, self.y / mag, self.z / mag)
    }
}

impl Vec4 {
    pub fn new(x: f32, y: f32, z: f32, w: f32) -> Vec4 {
        Vec4 { x, y, z, w }
    }

    pub fn zero() -> Vec4 {
        Vec4 {
            x: 0.0,
            y: 0.0,
            z: 0.0,
            w: 0.0,
        }
    }
}

impl Vertex {
    pub fn new(pos: Vec3, tex_coord: Vec2) -> Vertex {
        Vertex { pos, tex_coord }
    }
}

// column major ordering reference
// 0, 4, 8, 12,
// 1, 5, 9, 13,
// 2, 6, 10, 14,
// 3, 7, 11, 15
impl Mat4 {
    pub fn identity() -> Mat4 {
        Mat4 {
            data: [
                1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0,
            ],
        }
    }

    pub fn translation(vec: Vec3) -> Mat4 {
        let mut mat = Mat4::identity();
        mat.data[12] = vec.x;
        mat.data[13] = vec.y;
        mat.data[14] = vec.z;
        mat
    }

    pub fn translate(self, vec: Vec3) -> Mat4 {
        self * Mat4::translation(vec)
    }

    pub fn ortho(l: f32, r: f32, t: f32, b: f32, f: f32, n: f32) -> Mat4 {
        Mat4 {
            data: [
                2.0 / (r - l),
                0.0,
                0.0,
                0.0,
                0.0,
                2.0 / (t - b),
                0.0,
                0.0,
                0.0,
                0.0,
                -2.0 / (f - n),
                0.0,
                -(r + l) / (r - l),
                -(t + b) / (t - b),
                -(f + n) / (f - n),
                1.0,
            ],
        }
    }
}

impl std::ops::Mul for Mat4 {
    type Output = Self;

    fn mul(self, rhs: Self) -> Self {
        let mut res = Mat4::identity();
        for i in 0..4 {
            for j in 0..4 {
                res.data[i + (4 * j)] = (self.data[i] * rhs.data[j * 4])
                    + (self.data[i + 4] * rhs.data[(j * 4) + 1])
                    + (self.data[i + 8] * rhs.data[(j * 4) + 2])
                    + (self.data[i + 12] * rhs.data[(j * 4) + 3]);
            }
        }

        res
    }
}

impl Rect {
    pub fn new(pos: Vec2, w: f32, h: f32) -> Rect {
        Rect { pos, w, h }
    }

    pub fn from_parts(x: f32, y: f32, w: f32, h: f32) -> Rect {
        Rect::new(Vec2::new(x, y), w, h)
    }

    fn overlaps(&self, other: &Rect) -> bool {
        // flipping arguments captures containment
        overlaps(&self, other) || overlaps(other, &self)
    }
}

// determine if a rectangle overlaps another rectangle
fn overlaps(lhs: &Rect, rhs: &Rect) -> bool {
    let lhs_top = lhs.pos.y;
    let lhs_bottom = lhs.pos.y + lhs.h;
    let lhs_left = lhs.pos.x;
    let lhs_right = lhs.pos.x + lhs.w;

    let rhs_top = rhs.pos.y;
    let rhs_bottom = rhs.pos.y + rhs.h;
    let rhs_left = rhs.pos.x;
    let rhs_right = rhs.pos.x + rhs.w;

    let y_overlap = (lhs_top >= rhs_top && lhs_top <= rhs_bottom)
        || (lhs_bottom >= rhs_top && lhs_bottom <= rhs_bottom);
    let x_overlap = (lhs_left >= rhs_left && lhs_left <= rhs_right)
        || (lhs_right >= rhs_left && lhs_right <= rhs_right);

    return x_overlap && y_overlap;
}

impl From<&Vec2> for Vec3 {
    fn from(v: &Vec2) -> Vec3 {
        Vec3::new(v.x, v.y, 0.0)
    }
}

impl From<&Vec3> for Vec4 {
    fn from(v: &Vec3) -> Vec4 {
        Vec4::new(v.x, v.y, v.z, 0.0)
    }
}

fn f32_eq(lhs: f32, rhs: f32) -> bool {
    (lhs - rhs) < 0.0000001
}
